pkg/networkservice/common/connect/server.go

```diff
@@ -65,80 +97,204 @@ func WithDialOptionFactory(factory func(ctx context.Context, request *networkser
+		}
 	}
 	return &connectServer{
-		ctx:                     nil,
-		clientFactory:           clientFactory,
-		uRLToClientMap:          newClientMap(),
-		connectionIDToClientMap: newClientMap(),
-		dialOptions:             clientDialOptions,
-		dialOptionFactories:     dialOptionFactories,
+		ctx:                 nil,
+		clientFactory:       clientFactory,
+		clientMap:           map[string]*clientEntry{},
+		connections:         map[string]*clientEntry{},
+		dialOptions:         clientDialOptions,
+		dialOptionFactories: dialOptionFactories,
+		dialer:              dialer,
 	}
 }
 
-func (c *connectServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
-	client, found := c.connectionIDToClientMap.Load(request.GetConnection().GetId())
-	if found {
-		return client.Request(ctx, request)
-	}
-	u := clienturl.ClientURL(ctx)
-	client, found = c.uRLToClientMap.Load(u.String())
-	if found {
-		c.connectionIDToClientMap.LoadOrStore(request.GetConnection().GetId(), client)
-		return client.Request(ctx, request)
-	}
-	// TODO - do something with cancelFunc to clean up uRLToClientMap when we are done with them (maybe ref counting on Close?)
-	clientCtx, _ := context.WithCancel(context.Background())
-	// TODO - fix to accept dialOptions from github.com/networkservicemesh/sdk/tools/security - the options should flow in from the top
-	// TODO - fix to accept a mockable interface we can pass in from the top other than DialContext
-	// TODO - fix to be cautious about schemes
-	dialOptions := c.dialOptions
-	for _, dof := range c.dialOptionFactories {
-		dialOptions = append(dialOptions, dof(ctx, request, u)...)
+func grpcDialer(ctx context.Context, target string, opts ...grpc.DialOption) (conn grpc.ClientConnInterface, cancelFunc func(), err error) {
+	var grpcCon *grpc.ClientConn
+	grpcCon, err = grpc.DialContext(ctx, target, opts...)
+	cancelFunc = func() {
+		if grpcCon != nil {
+			_ = grpcCon.Close()
+		}
 	}
-	cc, err := grpc.DialContext(clientCtx, grpcutils.URLToTarget(u), dialOptions...)
+	return grpcCon, cancelFunc, err
+}
+
+func (c *connectServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+	// Find or Create clientEntry, with add pending.
+	ce, deferFunc, err := c.findOrCreateClient(ctx)
 	if err != nil {
-		return nil, errors.Wrapf(err, "Unable to dial %s", u.String())
+		// In case of error, no pending state is propogated.
+		return nil, err
 	}
-	client = c.clientFactory(clientCtx, cc)
-	client, _ = c.uRLToClientMap.LoadOrStore(u.String(), client)
-	client, _ = c.connectionIDToClientMap.LoadOrStore(request.GetConnection().GetId(), client)
-	if _, err = client.Request(ctx, request); err != nil {
+	// Will decrease pending and close connection if required
+	defer deferFunc()
+
+	// We have connection ready,
+	var conn *networkservice.Connection
+	conn, err = ce.client.Request(ctx, request)
+	if err != nil {
```

Comment:
Question: I am pretty sure that grpc redials itself if there's an issue (ie, it doesn't fail forever).  What circumstances do you expect in which we would need to reconstitute the whole client?

Reply:
Not sure I understand correctly, grpc created not here, but at line 281, this is handling of client response, and if it failed we need to remove connection stuff. For grpc I've expect only initial dial could fail.

Reply:
My concern is that we wind up removing the client no matter what the failure is... it could be as simple as a policy rejection of that specific Request...

Reply:
Agree, let's cleanup on close only.